### DISCUSSION TITLE
Initial observations/questions

### DISCUSSION THREAD

**Discussion by Dennis4b** [cloudnative-pg/cloudnative-pg#1465] [General]
Hi,

I am evaluating cloudnativepg to replace a manually managed etcd+patroni+postgresql environment, have played with a bare `cnpg` in `k3s`, and have some questions (though the documentation is very thorough!).

This is based on running the basic 3-instance example, except I use `minSyncReplicas: 1` and `maxSyncReplicas: 1`. There is almost no data inside so filesizes etc should be small.

* When I call `kubectl cnpg promote cluster-example cluster-example-1`, the new master is seen very fast. Soon-ish, the first replica will also be `Standby (sync)`. However, usually the second replica will be in `Standby (file based)` and can take literally minutes to achieve `Standby (sync)` state. I confirmed this by looking inside this replica and indeed not seeing any changes from the primary propagated. I would expect, for such a small database, that both replicas should be available very quickly (seconds).
Is this normal? (and why?) or do I have some network misconfiguration/timeouts or such? (I don't have an external WAL archive yet but this should not be needed as long as the replicas come back quickly enough that the primary has not removed its archives yet?)

* I don't really find any documentation regarding use of `pg_rewind`. I want to always have at least one synchronous replica, and a failover should never automatically rewind a timeline (= where this means lose data). I assume failover is always done to the replica which is up to date (there should be always at least one)? (in Patroni this is an explicit option)

* Did I read the documentation correctly, in that if both replicas would be down, `cnpg` would allow the remaining instance to run as a primary without any replication, regardless of `minSyncReplicas etc` settings? (this is probably what I would want)

* While I like the `synchronous_standby_names=ANY 1 (2,3)`, this means you can not use synchronous standby replicas as fully up-to-date read-only instances, as you're not guaranteed any particular replica is the fully up-to-date one. My app has all its queries annotated as read-write and read-only, but for that a read only replica should be always up to date (and with `synchronous_commit = remote_apply`). How would this work? If it is possible to configure one replica as always fully sync (instead of `ANY`), you would need an endpoint to expose which one this is? 

* I could change my app to loosen the restriction of `remote_apply` to use read-only databases that are slightly behind for the 95% of queries where this is probably fine, so both replicas could be used, but I would still like to limit the lag, i.e. if a replica spends 3 minutes in `Standby (file based)` I don't want to run queries on it as that is too much lag.

* If I understand the `pgbouncer` sections correctly, this removes the need in my app to deal with DNS updates/caching resolving the `-rw` endpoint continuously, but rather use a fixed `pgbouncer` `-rw` endpoint which is automatically directed at the primary including after promotation/failover?

Thank you!
Dennis

**Comment by gbartolini** [comment]
Hi Dennis,

> I am evaluating cloudnativepg to replace a manually managed etcd+patroni+postgresql environment, have played with a bare `cnpg` in `k3s`, and have some questions (though the documentation is very thorough!).

Awesome.

> * When I call `kubectl cnpg promote cluster-example cluster-example-1`, the new master is seen very fast. Soon-ish, the first replica will also be `Standby (sync)`. However, usually the second replica will be in `Standby (file based)` and can take literally minutes to achieve `Standby (sync)` state. I confirmed this by looking inside this replica and indeed not seeing any changes from the primary propagated. I would expect, for such a small database, that both replicas should be available very quickly (seconds).
>   Is this normal? (and why?) or do I have some network misconfiguration/timeouts or such? (I don't have an external WAL archive yet but this should not be needed as long as the replicas come back quickly enough that the primary has not removed its archives yet?)

If you don't have a WAL archive, my advice is to configure replication slots for HA as that should prevent this case from happening. It is suspicious you have file based standby without WAL archive.

> * I don't really find any documentation regarding use of `pg_rewind`. I want to always have at least one synchronous replica, and a failover should never automatically rewind a timeline (= where this means lose data). I assume failover is always done to the replica which is up to date (there should be always at least one)? (in Patroni this is an explicit option)

pg_rewind is used always, even if it does not do anything in case the former primary is aligned. At the moment, we are favouring self-healing over anything else, hence why pg_rewind takes precedence (and we cannot exclude data loss scenarios). In any case, in production, this would be treated as a PITR scenario. And you should have a backup in place, with WAL archiving. My advice is to have also a separate WAL volume and replication slots as anticipated above.

> * Did I read the documentation correctly, in that if both replicas would be down, `cnpg` would allow the remaining instance to run as a primary without any replication, regardless of `minSyncReplicas etc` settings? (this is probably what I would want)

Correct, we override the min setting to the number of available replicas (if 0, we set it to 0).

> * While I like the `synchronous_standby_names=ANY 1 (2,3)`, this means you can not use synchronous standby replicas as fully up-to-date read-only instances, as you're not guaranteed any particular replica is the fully up-to-date one. My app has all its queries annotated as read-write and read-only, but for that a read only replica should be always up to date (and with `synchronous_commit = remote_apply`). How would this work? If it is possible to configure one replica as always fully sync (instead of `ANY`), you would need an endpoint to expose which one this is?

This is a fair point. At the moment you need to treat them all at the same level and set remote apply as sync commit (as you say below).

> * I could change my app to loosen the restriction of `remote_apply` to use read-only databases that are slightly behind for the 95% of queries where this is probably fine, so both replicas could be used, but I would still like to limit the lag, i.e. if a replica spends 3 minutes in `Standby (file based)` I don't want to run queries on it as that is too much lag.

This is an anomaly. Could you help us with a reproducible use case?

> * If I understand the `pgbouncer` sections correctly, this removes the need in my app to deal with DNS updates/caching resolving the `-rw` endpoint continuously, but rather use a fixed `pgbouncer` `-rw` endpoint which is automatically directed at the primary including after promotation/failover?

Correct.

Ciao,
Gabriele

**Reply by Dennis4b** [reply]
Thank you for your response Gabriele.

My configuration was basically just the quickstart example (which does not enable HA replication slots nor WAL storage) with synchronous replication added, just to get a quick feel for things, though with running my 3 instances each on one of the three k8s master nodes with embedded etcd, so once I kill the network (to simulate failure) it simultaneously degrades the etcd cluster and the control plane.

I'm still getting comfortable with k8s itself so I need to do a lot more experimenting still. For now let's assume I did something wrong :)

As for the lagging replica part, I suppose in theory any node hosting a replica could suffer from a degraded network or disk, causing replication to be unable to keep up with the primary and falling further and further behind.

This can be a problem if the application uses replicas for read-only queries as there is no guarantee how fresh the data is.

I believe Patroni explicitly monitors how far behind a replica is and will take it out of the healthy set if it is too far behind (= configurable amount). I was wondering what CNPG's approach/opinion about this case is, or if this is handled already somehow.

Grazie!
